Property group reordering

Alex--C opened the pull request:
<!--
Thank you for contributing to Shopware! Please fill out this description template to help us to process your pull request.

Please make sure to fulfil our contribution guideline (https://docs.shopware.com/en/shopware-platform-dev-en/community/contribution-guideline?category=shopware-platform-dev-en/community).

Do your changes need to be mentioned in the documentation?
Add notes on your change right now in the documentation files in /src/Docs/Resources and add them to the pull request as well. 
-->

### 1. Why is this change necessary?

Options already support custom ordering using either a "position" field or by ordering them in the "Storefront presentation" modal on a per-product level. Property groups however do not support custom ordering, they are always ordered alphabetically in the configurator.

### 2. What does this change do, exactly?

If a custom order was set through the "Storefront presentation" modal for the product, the property groups will be sorted according to that order.

A new `position` field on the `PropertyGroup` (similar to `PropertyGroupOption`) is used as a fallback if no product-level order is specified for the product. By sorting `PropertyGroup` with the same `position` by their translated name, the default behaviour will be backwards compatible. So if no `position` is set, the ordering will be the same as without this feature.

The property filters are also sorted according to the `position` field, as it makes no sense to use product-specific ordering there (the displayed products might be using different property group orders).

### 3. Describe each step to reproduce the issue or behaviour.

- Create or edit a PropertyGroup
- Assign a `position` to the group or change the order on product-level
- Create variations
- The `PropertyGroup`s in the configurator will be ordered correctly

### 4. Please link to the relevant issues (if any).


### 5. Checklist

- [x] I have written tests and verified that they fail without my change
- [x] I have squashed any insignificant commits
- [ ] I have written or adjusted the documentation according to my changes
- [x] This change has comments for package types, values, functions, and non-obvious lines of code
- [x] I have read the contribution requirements and fulfil them.

Alex--C committed:
Implement property group reordering

hlohaus commented on src/Core/Migration/Migration1578590702AddedPropertyGroupPosition.php:
```diff
@@ -0,0 +1,24 @@
+
+    public function updateDestructive(Connection $connection): void
+    {
+        $connection->executeUpdate('ALTER TABLE `property_group_translation` DROP COLUMN `position`;');
```
that's not meant for that

JoshuaBehrens replied:
Destructive updates are updates that execute irreversible schema changes. The idea is on a normal update you just do stuff that is compatible with older code changes. Any destructive change is about to break previous code. This is called blue-green deployment.

Alex--C replied:
... I'm so used to Symfony's up/down methods in the migrations that I didn't even read the method names. Thanks, will fix!

JoshuaBehrens commented on src/Storefront/Page/Product/Configurator/ProductPageConfiguratorLoader.php:
```diff
@@ -154,7 +154,12 @@ private function sortSettings(array $groups): PropertyGroupCollection
+                if ($posA === $posB) {
+                    return strnatcmp($a->getTranslation('name'), $b->getTranslation('name'));
+                }
+                return ($posA < $posB) ? -1 : 1;
```
As of php 7.0 there is a spaceship operator. 
`return $posA <=> $posB;`

Alex--C replied:
TIL, thanks.

OliverSkroblin commented:
Hi @Alex--C,

we already implemented this feature. The groups can be sorted for each product. The config for are stored inside the `configurator_group_config` field in the product table.

Alex--C commented:
Hi @OliverSkroblin,

Thanks for your response. I'm not sure whether we are talking about the same thing. To me it seems like the groups are ordered by their translated name in https://github.com/shopware/platform/blob/master/src/Storefront/Page/Product/Configurator/ProductPageConfiguratorLoader.php#L154 . That result seems to go directly into the product configurator templates without further sorting.

How can one change the order of the groups for the configurator using the field you mentioned?

Alex--C commented:
@OliverSkroblin I did some more digging and it seems like the admin provides the functionality to add custom ordering to the groups, but the storefront simply ignores this order when generating the configurator.
I extended my PR, it now sorts the configurator property groups according to the product's `configurator_group_config` field. If that is empty, it falls back to my new `position` field, which allows for "global" sorting configuration if product-specific ordering is not needed.

Alex--C committed:
Consider configuratorGroupConfig in configurator group sorting

Alex--C committed:
Add property filter sorting

Alex--C commented:
@OliverSkroblin any news regarding this topic? If you are planning to implement something like this on your own it would be cool if you could give some sort of time horizon - one of our projects depends on this functionality and we need to know whether it will be in the core or if we need to hack it with a plugin.

OliverSkroblin commented:
Hi @Alex--C, 

Sorry for the late reply. We had a little bit to do with the final release of v6.1.0 ;).

Basically the idea is good. I'll talk it over with my product owner, if we want to implement the position at the groups.

I will left some more comments in the PR that still need to be corrected.

OliverSkroblin commented on src/Core/Content/Property/Aggregate/PropertyGroupTranslation/PropertyGroupTranslationDefinition.php:
```diff
@@ -39,6 +40,7 @@ protected function defineFields(): FieldCollection
         return new FieldCollection([
             (new StringField('name', 'name'))->addFlags(new Required()),
             new LongTextField('description', 'description'),
+            new IntField('position', 'position'),
```
please define a default in the getDefaults function.

Alex--C replied:
Implemented getDefaults.

OliverSkroblin commented on src/Core/Content/Property/Aggregate/PropertyGroupTranslation/PropertyGroupTranslationEntity.php:
```diff
@@ -22,6 +22,11 @@ class PropertyGroupTranslationEntity extends TranslationEntity
+    /**
+     * @var int|null
+     */
+    protected $position;
```
position should not be nullable, please add a database and definition default.

Alex--C replied:
getDefaults and database default value added.

OliverSkroblin commented on src/Core/Content/Property/PropertyGroupEntity.php:
```diff
@@ -86,6 +91,16 @@ public function setDescription(?string $description): void
         $this->description = $description;
     }
 
+    public function getPosition(): ?int
```
After adding a default to storage and definition, this value should be always an integer

Alex--C replied:
Defaults added, removed nullable.

OliverSkroblin commented on src/Storefront/Page/Product/Configurator/ProductPageConfiguratorLoader.php:
```diff
@@ -179,7 +188,30 @@ static function (PropertyGroupOptionEntity $a, PropertyGroupOptionEntity $b) use
+            ->search($criteria, $context->getContext())
+            ->first();
+
+        $configuratorGroupConfig = $parentProduct->getConfiguratorGroupConfig();
```
If you add the Inherited flag to this field, you don't have to fetch the parent at this point. 
The configuration is only saved to the parent product.

OliverSkroblin replied:
A few tests for this change would also be important.

Alex--C replied:
Added Inherited field, added test for property group order.

OliverSkroblin commented:
I spoke to my product owner and he likes the idea of the position for a group. If you make the remaining adjustments, I would merge the PR. The changes would be available with v6.2.0.

Alex--C commented:
Awesome news, thanks!
Also thank you for the code review. I'll make the changes this afternoon.

mitelg commented:
@Alex--C if we are going for 6.2, it would be nice, if you set the target branch to "master", as this would be the future 6.2 version :wink:

Alex--C committed:
Implement code review feedback and test

Alex--C commented:
@OliverSkroblin I implemented the changes and added a test for the group ordering, I hope the way I implemented it makes sense. 

@mitelg Target changed to master :-)

JoshuaBehrens commented:
🎉 Tests

mitelg commented:
your PR shows me, that there is currently a misconfiguration in the travis pipeline :grin: the cs-fixer should have complained about some of your code :wink: 
I will fix that

OliverSkroblin commented:
Fixed with: https://github.com/shopware/platform/commit/ecf996177b624fa3635e1b0033ff9a0dfa7a7315
